examples/which-problem.kl

```diff
@@ -28,13 +30,11 @@
                    [(expression t)
                     (type-case t
                       [(-> a b)
-                       (pure (list-syntax ('lambda
-                                           (list-syntax ('_) stx)
-                                           (list-syntax ('mega-const e) e))
-                                          stx))]
+                       (pure `(lambda (_)
+                                (mega-const ,e)))]
                       [(else x) (pure e)])])))]))]))
 
-(example (the (-> (Bool) (-> (Bool) (-> (Bool) (-> (Bool) (Unit))))) (mega-const (unit))))
+(example (the (-> (Bool) (Bool) (Bool) (Bool) (Unit)) (mega-const (unit))))
```

Comment:
I think it would look even nicer if there weren't parentheses around `Bool`, `Unit` and `unit`, is there a particular reason for having them?

Reply:
Yes - it's to have a uniform syntax for type construction, especially once we get higher-kinded stuff. The idea is that the core language is as simple and uniform as possible, and then we can define macros to make it prettier.  FWIW, this is one of the reasons for having patterns and type patterns be macro expansion positions.

Try something like this:
```klister
(define-macros
  ([Unit* (lambda (stx) (syntax-case stx [(ident u) (quasiquote/loc stx Unit)]))]))
(export (rename Unit* Unit))
```